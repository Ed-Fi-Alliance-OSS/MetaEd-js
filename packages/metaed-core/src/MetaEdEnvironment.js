// @flow
import { newEntityRepository } from './model/EntityRepository';
import { newPropertyIndex } from './model/property/PropertyRepository';
import type { EntityRepository } from './model/EntityRepository';
import type { PluginEnvironment } from './plugin/PluginEnvironment';
import type { PropertyIndex } from './model/property/PropertyRepository';

export type SemVer = string;

// provided to validators and enhancers
export type MetaEdEnvironment = {
  // the entity repository
  entity: EntityRepository,

  // the property index, provided for convenience (a little hesitant here, as it should be dependent on entityRepository)
  propertyIndex: PropertyIndex,

  // plugin environment by plugin name
  plugin: Map<string, PluginEnvironment>,

  // the current MetaEd Version
  metaEdVersion: SemVer,

  // the data standard version to target
  dataStandardVersion: SemVer,
};

export const newMetaEdEnvironment: () => MetaEdEnvironment = () => ({
  entity: newEntityRepository(),
  propertyIndex: newPropertyIndex(),
  plugin: new Map(),
  metaEdVersion: '0.0.0',
  dataStandardVersion: '0.0.0',
});
