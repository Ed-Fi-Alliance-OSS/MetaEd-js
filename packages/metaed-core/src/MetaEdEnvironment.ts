import { newPropertyIndex } from './model/property/PropertyRepository';
import { Namespace } from './model/Namespace';
import { PluginEnvironment } from './plugin/PluginEnvironment';
import { PropertyIndex } from './model/property/PropertyRepository';

/**
 *
 */
export type SemVer = string;

/**
 * Provided to validators and enhancers
 */
export type MetaEdEnvironment = {
  // namespaces
  namespace: Map<string, Namespace>;

  // the property index, provided for convenience (a little hesitant here, as it should be dependent on entityRepository)
  propertyIndex: PropertyIndex;

  // plugin environment by plugin shortName
  plugin: Map<string, PluginEnvironment>;

  // the current MetaEd Version
  metaEdVersion: SemVer;

  // the data standard version to target
  dataStandardVersion: SemVer;
};

/**
 *
 */
export const newMetaEdEnvironment: () => MetaEdEnvironment = () => ({
  namespace: new Map(),
  propertyIndex: newPropertyIndex(),
  plugin: new Map(),
  metaEdVersion: '0.0.0',
  dataStandardVersion: '0.0.0',
});
