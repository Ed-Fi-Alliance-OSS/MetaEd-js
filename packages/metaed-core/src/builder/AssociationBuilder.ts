import { MetaEdGrammar } from '../grammar/gen/MetaEdGrammar';
import { TopLevelEntityBuilder } from './TopLevelEntityBuilder';
import { newDomainEntityProperty } from '../model/property/DomainEntityProperty';
import { newAssociation } from '../model/Association';
import { NoTopLevelEntity } from '../model/TopLevelEntity';
import { isErrorText } from './BuilderUtility';
import { sourceMapFrom } from '../model/SourceMap';

/**
 * An ANTLR4 listener that creates Association entities.
 */
export class AssociationBuilder extends TopLevelEntityBuilder {
  enterAssociation(context: MetaEdGrammar.AssociationContext) {
    this.enteringEntity(newAssociation);
    if (this.currentTopLevelEntity !== NoTopLevelEntity) {
      this.currentTopLevelEntity.sourceMap.type = sourceMapFrom(context);
    }
  }

  // @ts-ignore
  exitAssociation(context: MetaEdGrammar.AssociationContext) {
    this.exitingEntity();
  }

  enterAssociationName(context: MetaEdGrammar.AssociationNameContext) {
    if (
      this.currentTopLevelEntity === NoTopLevelEntity ||
      context.exception ||
      context.ID() == null ||
      context.ID().exception ||
      isErrorText(context.ID().getText())
    )
      return;
    this.enteringName(context.ID().getText());
    this.currentTopLevelEntity.sourceMap.metaEdName = sourceMapFrom(context);
  }

  enterCascadeUpdate(context: MetaEdGrammar.CascadeUpdateContext) {
    if (this.currentTopLevelEntity !== NoTopLevelEntity) {
      this.currentTopLevelEntity.allowPrimaryKeyUpdates = true;
      this.currentTopLevelEntity.sourceMap.allowPrimaryKeyUpdates = sourceMapFrom(context);
    }
  }

  enterFirstDomainEntity(context: MetaEdGrammar.FirstDomainEntityContext) {
    if (this.currentTopLevelEntity === NoTopLevelEntity) return;
    if (context.exception) return;
    this.currentProperty = newDomainEntityProperty();
    this.currentProperty.sourceMap.type = sourceMapFrom(context);
    this.enteringIdentity(context);
  }

  enterSecondDomainEntity(context: MetaEdGrammar.SecondDomainEntityContext) {
    if (this.currentTopLevelEntity === NoTopLevelEntity) return;
    if (context.exception) return;
    this.currentProperty = newDomainEntityProperty();
    this.currentProperty.sourceMap.type = sourceMapFrom(context);
    this.enteringIdentity(context);
  }

  // @ts-ignore
  exitFirstDomainEntity(context: MetaEdGrammar.FirstDomainEntityContext) {
    this.exitingProperty();
  }

  // @ts-ignore
  exitSecondDomainEntity(context: MetaEdGrammar.SecondDomainEntityContext) {
    this.exitingProperty();
  }
}
