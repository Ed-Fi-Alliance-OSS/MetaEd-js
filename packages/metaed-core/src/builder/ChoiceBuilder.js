// @flow
import type { MetaEdGrammar } from '../grammar/gen/MetaEdGrammar';
import { TopLevelEntityBuilder } from './TopLevelEntityBuilder';
import { newChoice } from '../model/Choice';
import { isErrorText } from './BuilderUtility';
import { NoTopLevelEntity } from '../model/TopLevelEntity';
import { sourceMapFrom } from '../model/SourceMap';

export class ChoiceBuilder extends TopLevelEntityBuilder {
  enterChoice(context: MetaEdGrammar.ChoiceContext) {
    this.enteringEntity(newChoice);
    if (this.currentTopLevelEntity !== NoTopLevelEntity) {
      this.currentTopLevelEntity.sourceMap.type = sourceMapFrom(context);
    }
  }

  // eslint-disable-next-line no-unused-vars
  exitChoice(context: MetaEdGrammar.ChoiceContext) {
    this.exitingEntity();
  }

  enterChoiceName(context: MetaEdGrammar.ChoiceNameContext) {
    if (this.currentTopLevelEntity === NoTopLevelEntity) return;
    if (context.exception || context.ID() == null || context.ID().exception || isErrorText(context.ID().getText())) return;
    this.enteringName(context.ID().getText());
    this.currentTopLevelEntity.sourceMap.metaEdName = sourceMapFrom(context);
  }
}
