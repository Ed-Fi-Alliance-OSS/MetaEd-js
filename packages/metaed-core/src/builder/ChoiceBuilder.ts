import type { ChoiceContext, ChoiceNameContext } from '../grammar/gen/MetaEdGrammar';
import { TopLevelEntityBuilder } from './TopLevelEntityBuilder';
import { newChoice } from '../model/Choice';
import { isErrorText } from './BuilderUtility';
import { NoTopLevelEntity } from '../model/TopLevelEntity';
import { sourceMapFrom } from '../model/SourceMap';

/**
 * An ANTLR4 listener that creates Choice entities.
 */
export class ChoiceBuilder extends TopLevelEntityBuilder {
  enterChoice = (context: ChoiceContext) => {
    this.enteringEntity(newChoice);
    if (this.currentTopLevelEntity !== NoTopLevelEntity) {
      this.currentTopLevelEntity.sourceMap.type = sourceMapFrom(context);
    }
  };

  exitChoice = (_context: ChoiceContext) => {
    this.exitingEntity();
  };

  enterChoiceName = (context: ChoiceNameContext) => {
    if (this.currentTopLevelEntity === NoTopLevelEntity) return;
    if (context.exception || context.ID() == null || isErrorText(context.ID().getText())) return;
    this.enteringName(context.ID().getText());
    this.currentTopLevelEntity.sourceMap.metaEdName = sourceMapFrom(context);
  };
}
