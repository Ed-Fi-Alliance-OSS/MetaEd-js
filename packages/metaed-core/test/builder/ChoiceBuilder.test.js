// @flow
import { ChoiceBuilder } from '../../src/builder/ChoiceBuilder';
import { NamespaceInfoBuilder } from '../../src/builder/NamespaceInfoBuilder';
import { MetaEdTextBuilder } from '../../src/grammar/MetaEdTextBuilder';
import { newMetaEdEnvironment } from '../../src/MetaEdEnvironment';
import { getChoice } from '../TestHelper';
import type { MetaEdEnvironment } from '../../src/MetaEdEnvironment';
import type { ValidationFailure } from '../../src/validator/ValidationFailure';

describe('when building choice in extension namespace', () => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const validationFailures: Array<ValidationFailure> = [];
  const namespace: string = 'namespace';
  const projectExtension: string = 'ProjectExtension';

  const entityName: string = 'EntityName';
  const entityMetaEdId: string = '1';
  const propertyName: string = 'PropertyName';
  const propertyMetaEdId: string = '2';
  const propertyDocumentation: string = 'PropertyDocumentation';
  const entityDocumentation: string = 'EntityDocumentation';

  beforeAll(() => {
    const builder = new ChoiceBuilder(metaEd, validationFailures);

    MetaEdTextBuilder.build()
      .withBeginNamespace(namespace, projectExtension)
      .withStartChoice(entityName, entityMetaEdId)
      .withDocumentation(entityDocumentation)
      .withIntegerProperty(propertyName, propertyDocumentation, true, false, null, null, null, propertyMetaEdId)
      .withEndChoice()
      .withEndNamespace()
      .sendToListener(new NamespaceInfoBuilder(metaEd, validationFailures))
      .sendToListener(builder);
  });

  it('should build one choice', () => {
    expect(metaEd.entity.choice.size).toBe(1);
  });

  it('should be found in entity repository', () => {
    expect(getChoice(metaEd.entity, entityName)).toBeDefined();
    expect(getChoice(metaEd.entity, entityName).metaEdName).toBe(entityName);
  });

  it('should have no validation failures', () => {
    expect(validationFailures).toHaveLength(0);
  });

  it('should have namespace', () => {
    expect(getChoice(metaEd.entity, entityName).namespaceInfo.namespace).toBe(namespace);
  });

  it('should have metaEdId', () => {
    expect(getChoice(metaEd.entity, entityName).metaEdId).toBe(entityMetaEdId);
  });

  it('should have project extension', () => {
    expect(getChoice(metaEd.entity, entityName).namespaceInfo.projectExtension).toBe(projectExtension);
  });

  it('should have documentation', () => {
    expect(getChoice(metaEd.entity, entityName).documentation).toBe(entityDocumentation);
  });

  it('should have one property', () => {
    expect(getChoice(metaEd.entity, entityName).properties).toHaveLength(1);
  });

  it('should have integer property', () => {
    const property = getChoice(metaEd.entity, entityName).properties[0];

    expect(property.metaEdName).toBe(propertyName);
    expect(property.type).toBe('integer');
    expect(property.metaEdId).toBe(propertyMetaEdId);
    expect(property.isRequired).toBe(true);
    expect(property.documentation).toBe(propertyDocumentation);
  });
});

describe('when building duplicate choices', () => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const validationFailures: Array<ValidationFailure> = [];
  const namespace: string = 'namespace';
  const projectExtension: string = 'ProjectExtension';

  const entityName: string = 'EntityName';
  const entityMetaEdId: string = '1';
  const propertyName: string = 'PropertyName';
  const propertyMetaEdId: string = '2';
  const propertyDocumentation: string = 'PropertyDocumentation';
  const entityDocumentation: string = 'EntityDocumentation';

  beforeAll(() => {
    const builder = new ChoiceBuilder(metaEd, validationFailures);

    MetaEdTextBuilder.build()
      .withBeginNamespace(namespace, projectExtension)
      .withStartChoice(entityName, entityMetaEdId)
      .withDocumentation(entityDocumentation)
      .withIntegerProperty(propertyName, propertyDocumentation, true, false, null, null, null, propertyMetaEdId)
      .withEndChoice()

      .withStartChoice(entityName, entityMetaEdId)
      .withDocumentation(entityDocumentation)
      .withIntegerProperty(propertyName, propertyDocumentation, true, false, null, null, null, propertyMetaEdId)
      .withEndChoice()
      .withEndNamespace()
      .sendToListener(new NamespaceInfoBuilder(metaEd, validationFailures))
      .sendToListener(builder);
  });

  it('should build one choice', () => {
    expect(metaEd.entity.choice.size).toBe(1);
  });

  it('should be found in entity repository', () => {
    expect(getChoice(metaEd.entity, entityName)).toBeDefined();
    expect(getChoice(metaEd.entity, entityName).metaEdName).toBe(entityName);
  });

  it('should have two validation failures', () => {
    expect(validationFailures).toHaveLength(2);
  });

  it('should have validation failures for each entity', () => {
    expect(validationFailures[0].validatorName).toBe('TopLevelEntityBuilder');
    expect(validationFailures[0].category).toBe('error');
    expect(validationFailures[0].message).toMatchSnapshot(
      'when building duplicate choices should have validation failures for each entity -> Choice 1 message',
    );
    expect(validationFailures[0].sourceMap).toMatchSnapshot(
      'when building duplicate choices should have validation failures for each entity -> Choice 1 sourceMap',
    );

    expect(validationFailures[1].validatorName).toBe('TopLevelEntityBuilder');
    expect(validationFailures[1].category).toBe('error');
    expect(validationFailures[1].message).toMatchSnapshot(
      'when building duplicate choices should have validation failures for each entity -> Choice 2 message',
    );
    expect(validationFailures[1].sourceMap).toMatchSnapshot(
      'when building duplicate choices should have validation failures for each entity -> Choice 2 sourceMap',
    );
  });
});

describe('when building choice with no choice name', () => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const validationFailures: Array<ValidationFailure> = [];
  const textBuilder: MetaEdTextBuilder = MetaEdTextBuilder.build();
  const namespace: string = 'namespace';
  const projectExtension: string = 'ProjectExtension';

  const entityName: string = '';
  const entityMetaEdId: string = '1';
  const entityDocumentation: string = 'EntityDocumentation';
  const propertyName: string = 'PropertyName';
  const propertyDocumentation: string = 'PropertyDocumentation';
  const propertyMetaEdId: string = '2';

  beforeAll(() => {
    const builder = new ChoiceBuilder(metaEd, validationFailures);

    textBuilder
      .withBeginNamespace(namespace, projectExtension)
      .withStartChoice(entityName, entityMetaEdId)
      .withDocumentation(entityDocumentation)
      .withIntegerProperty(propertyName, propertyDocumentation, true, false, null, null, null, propertyMetaEdId)
      .withEndChoice()
      .withEndNamespace()
      .sendToListener(new NamespaceInfoBuilder(metaEd, validationFailures))
      .sendToListener(builder);
  });

  it('should not build choice', () => {
    expect(metaEd.entity.choice.size).toBe(0);
  });

  it('should have missing id error', () => {
    expect(textBuilder.errorMessages).toMatchSnapshot();
  });
});

describe('when building choice with lowercase choice name', () => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const validationFailures: Array<ValidationFailure> = [];
  const textBuilder: MetaEdTextBuilder = MetaEdTextBuilder.build();
  const namespace: string = 'namespace';
  const projectExtension: string = 'ProjectExtension';

  const entityName: string = 'entityName';
  const expectedName: string = 'Name';
  const entityMetaEdId: string = '1';
  const entityDocumentation: string = 'EntityDocumentation';
  const propertyName: string = 'PropertyName';
  const propertyDocumentation: string = 'PropertyDocumentation';
  const propertyMetaEdId: string = '2';

  beforeAll(() => {
    const builder = new ChoiceBuilder(metaEd, validationFailures);

    textBuilder
      .withBeginNamespace(namespace, projectExtension)
      .withStartChoice(entityName, entityMetaEdId)
      .withDocumentation(entityDocumentation)
      .withIntegerProperty(propertyName, propertyDocumentation, true, false, null, null, null, propertyMetaEdId)
      .withEndChoice()
      .withEndNamespace()
      .sendToListener(new NamespaceInfoBuilder(metaEd, validationFailures))
      .sendToListener(builder);
  });

  it('should build one choice', () => {
    expect(metaEd.entity.choice.size).toBe(1);
  });

  it('should be found in entity repository but with lowercase prefix ignored', () => {
    expect(getChoice(metaEd.entity, expectedName)).toBeDefined();
    expect(getChoice(metaEd.entity, expectedName).metaEdName).toBe(expectedName);
  });

  it('should have no validation failures', () => {
    expect(validationFailures).toHaveLength(0);
  });

  it('should have namespace', () => {
    expect(getChoice(metaEd.entity, expectedName).namespaceInfo.namespace).toBe(namespace);
  });

  it('should have metaEdId', () => {
    expect(getChoice(metaEd.entity, expectedName).metaEdId).toBe(entityMetaEdId);
  });

  it('should have project extension', () => {
    expect(getChoice(metaEd.entity, expectedName).namespaceInfo.projectExtension).toBe(projectExtension);
  });

  it('should have documentation', () => {
    expect(getChoice(metaEd.entity, expectedName).documentation).toBe(entityDocumentation);
  });

  it('should have one property', () => {
    expect(getChoice(metaEd.entity, expectedName).properties).toHaveLength(1);
  });

  it('should have integer property', () => {
    const property = getChoice(metaEd.entity, expectedName).properties[0];

    expect(property.metaEdName).toBe(propertyName);
    expect(property.type).toBe('integer');
    expect(property.metaEdId).toBe(propertyMetaEdId);
    expect(property.isRequired).toBe(true);
    expect(property.documentation).toBe(propertyDocumentation);
  });

  it('should have extraneous input error', () => {
    expect(textBuilder.errorMessages).toMatchSnapshot();
  });
});

describe('when building choice with no documentation', () => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const validationFailures: Array<ValidationFailure> = [];
  const textBuilder: MetaEdTextBuilder = MetaEdTextBuilder.build();
  const namespace: string = 'namespace';
  const projectExtension: string = 'ProjectExtension';

  const entityName: string = 'EntityName';
  const entityMetaEdId: string = '1';
  const propertyName: string = 'PropertyName';
  const propertyDocumentation: string = 'PropertyDocumentation';
  const propertyMetaEdId: string = '2';

  beforeAll(() => {
    const builder = new ChoiceBuilder(metaEd, validationFailures);

    textBuilder
      .withBeginNamespace(namespace, projectExtension)
      .withStartChoice(entityName, entityMetaEdId)
      .withIntegerProperty(propertyName, propertyDocumentation, true, false, null, null, null, propertyMetaEdId)
      .withEndChoice()
      .withEndNamespace()
      .sendToListener(new NamespaceInfoBuilder(metaEd, validationFailures))
      .sendToListener(builder);
  });

  it('should be found in entity repository', () => {
    expect(getChoice(metaEd.entity, entityName)).toBeDefined();
    expect(getChoice(metaEd.entity, entityName).metaEdName).toBe(entityName);
  });

  it('should have no validation failures', () => {
    expect(validationFailures).toHaveLength(0);
  });

  it('should have namespace', () => {
    expect(getChoice(metaEd.entity, entityName).namespaceInfo.namespace).toBe(namespace);
  });

  it('should have metaEdId', () => {
    expect(getChoice(metaEd.entity, entityName).metaEdId).toBe(entityMetaEdId);
  });

  it('should have project extension', () => {
    expect(getChoice(metaEd.entity, entityName).namespaceInfo.projectExtension).toBe(projectExtension);
  });

  it('should have no documentation', () => {
    expect(getChoice(metaEd.entity, entityName).documentation).toBe('');
  });

  it('should have one property', () => {
    expect(getChoice(metaEd.entity, entityName).properties).toHaveLength(1);
  });

  it('should have integer property', () => {
    const property = getChoice(metaEd.entity, entityName).properties[0];

    expect(property.metaEdName).toBe(propertyName);
    expect(property.type).toBe('integer');
    expect(property.metaEdId).toBe(propertyMetaEdId);
    expect(property.isRequired).toBe(true);
    expect(property.documentation).toBe(propertyDocumentation);
  });

  it('should have mismatched input error', () => {
    expect(textBuilder.errorMessages).toMatchSnapshot();
  });
});

describe('when building choice with no property', () => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const validationFailures: Array<ValidationFailure> = [];
  const textBuilder: MetaEdTextBuilder = MetaEdTextBuilder.build();
  const namespace: string = 'namespace';
  const projectExtension: string = 'ProjectExtension';

  const entityName: string = 'EntityName';
  const entityMetaEdId: string = '1';
  const entityDocumentation: string = 'EntityDocumentation';

  beforeAll(() => {
    const builder = new ChoiceBuilder(metaEd, validationFailures);

    textBuilder
      .withBeginNamespace(namespace, projectExtension)
      .withStartChoice(entityName, entityMetaEdId)
      .withDocumentation(entityDocumentation)
      .withEndChoice()
      .withEndNamespace()
      .sendToListener(new NamespaceInfoBuilder(metaEd, validationFailures))
      .sendToListener(builder);
  });

  it('should be found in entity repository', () => {
    expect(getChoice(metaEd.entity, entityName)).toBeDefined();
    expect(getChoice(metaEd.entity, entityName).metaEdName).toBe(entityName);
  });

  it('should have no validation failures', () => {
    expect(validationFailures).toHaveLength(0);
  });

  it('should have namespace', () => {
    expect(getChoice(metaEd.entity, entityName).namespaceInfo.namespace).toBe(namespace);
  });

  it('should have metaEdId', () => {
    expect(getChoice(metaEd.entity, entityName).metaEdId).toBe(entityMetaEdId);
  });

  it('should have project extension', () => {
    expect(getChoice(metaEd.entity, entityName).namespaceInfo.projectExtension).toBe(projectExtension);
  });

  it('should have documentation', () => {
    expect(getChoice(metaEd.entity, entityName).documentation).toBe(entityDocumentation);
  });

  it('should have no property', () => {
    expect(getChoice(metaEd.entity, entityName).properties).toHaveLength(0);
  });

  it('should have mismatched input error', () => {
    expect(textBuilder.errorMessages).toMatchSnapshot();
  });
});

describe('when building choice with invalid trailing text', () => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const validationFailures: Array<ValidationFailure> = [];
  const textBuilder: MetaEdTextBuilder = MetaEdTextBuilder.build();
  const namespace: string = 'namespace';
  const projectExtension: string = 'ProjectExtension';

  const entityName: string = 'EntityName';
  const entityMetaEdId: string = '1';
  const entityDocumentation: string = 'EntityDocumentation';
  const propertyName: string = 'PropertyName';
  const propertyDocumentation: string = 'PropertyDocumentation';
  const propertyMetaEdId: string = '2';
  const trailingText: string = '\r\nTrailingText';

  beforeAll(() => {
    const builder = new ChoiceBuilder(metaEd, validationFailures);

    textBuilder
      .withBeginNamespace(namespace, projectExtension)
      .withStartChoice(entityName, entityMetaEdId)
      .withDocumentation(entityDocumentation)
      .withIntegerProperty(propertyName, propertyDocumentation, true, false, null, null, null, propertyMetaEdId)
      .withTrailingText(trailingText)
      .withEndChoice()
      .withEndNamespace()
      .sendToListener(new NamespaceInfoBuilder(metaEd, validationFailures))
      .sendToListener(builder);
  });

  it('should be found in entity repository', () => {
    expect(getChoice(metaEd.entity, entityName)).toBeDefined();
    expect(getChoice(metaEd.entity, entityName).metaEdName).toBe(entityName);
  });

  it('should have no validation failures', () => {
    expect(validationFailures).toHaveLength(0);
  });

  it('should have namespace', () => {
    expect(getChoice(metaEd.entity, entityName).namespaceInfo.namespace).toBe(namespace);
  });

  it('should have metaEdId', () => {
    expect(getChoice(metaEd.entity, entityName).metaEdId).toBe(entityMetaEdId);
  });

  it('should have project extension', () => {
    expect(getChoice(metaEd.entity, entityName).namespaceInfo.projectExtension).toBe(projectExtension);
  });

  it('should have no documentation', () => {
    expect(getChoice(metaEd.entity, entityName).documentation).toBe(entityDocumentation);
  });

  it('should have one property', () => {
    expect(getChoice(metaEd.entity, entityName).properties).toHaveLength(1);
  });

  it('should have integer property', () => {
    const property = getChoice(metaEd.entity, entityName).properties[0];

    expect(property.metaEdName).toBe(propertyName);
    expect(property.type).toBe('integer');
    expect(property.metaEdId).toBe(propertyMetaEdId);
    expect(property.isRequired).toBe(true);
    expect(property.documentation).toBe(propertyDocumentation);
  });

  it('should have extraneous input error', () => {
    expect(textBuilder.errorMessages).toMatchSnapshot();
  });
});

describe('when building choice source map', () => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const validationFailures: Array<ValidationFailure> = [];
  const namespace: string = 'namespace';
  const projectExtension: string = 'ProjectExtension';

  const entityName: string = 'EntityName';
  const entityMetaEdId: string = '1';
  const propertyName: string = 'PropertyName';
  const propertyMetaEdId: string = '2';
  const propertyDocumentation: string = 'PropertyDocumentation';
  const entityDocumentation: string = 'EntityDocumentation';

  beforeAll(() => {
    const builder = new ChoiceBuilder(metaEd, validationFailures);

    MetaEdTextBuilder.build()
      .withBeginNamespace(namespace, projectExtension)
      .withStartChoice(entityName, entityMetaEdId)
      .withDocumentation(entityDocumentation)
      .withIntegerProperty(propertyName, propertyDocumentation, true, false, null, null, null, propertyMetaEdId)
      .withEndChoice()
      .withEndNamespace()
      .sendToListener(new NamespaceInfoBuilder(metaEd, validationFailures))
      .sendToListener(builder);
  });

  it('should have a documentation property', () => {
    expect(getChoice(metaEd.entity, entityName).sourceMap.documentation).toBeDefined();
  });

  it('should have a metaEdId property', () => {
    expect(getChoice(metaEd.entity, entityName).sourceMap.metaEdId).toBeDefined();
  });

  it('should have a metaEdName property', () => {
    expect(getChoice(metaEd.entity, entityName).sourceMap.metaEdName).toBeDefined();
  });

  it('should have a type property', () => {
    expect(getChoice(metaEd.entity, entityName).sourceMap.type).toBeDefined();
  });

  it('should have source map data', () => {
    expect(getChoice(metaEd.entity, entityName).sourceMap).toMatchSnapshot();
  });
});
