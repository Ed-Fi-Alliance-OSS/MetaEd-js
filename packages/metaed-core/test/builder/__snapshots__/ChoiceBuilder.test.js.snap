// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when building choice source map should have source map data 1`] = `
Object {
  "allowPrimaryKeyUpdates": Object {
    "column": 0,
    "line": 0,
    "tokenText": "NoSourceMap",
  },
  "baseEntity": Object {
    "column": 0,
    "line": 0,
    "tokenText": "NoSourceMap",
  },
  "baseEntityName": Object {
    "column": 0,
    "line": 0,
    "tokenText": "NoSourceMap",
  },
  "documentation": Object {
    "column": 4,
    "line": 3,
    "tokenText": "documentation",
  },
  "identityProperties": Array [],
  "metaEdId": Object {
    "column": 20,
    "line": 2,
    "tokenText": "[1]",
  },
  "metaEdName": Object {
    "column": 9,
    "line": 2,
    "tokenText": "EntityName",
  },
  "properties": Array [],
  "queryableFields": Array [],
  "type": Object {
    "column": 2,
    "line": 2,
    "tokenText": "Choice",
  },
}
`;

exports[`when building choice with invalid trailing text should have extraneous input error 1`] = `
Array [
  "extraneous input 'TrailingText' expecting {'Abstract Entity', 'Association', 'End Namespace', 'Choice', 'Common', 'Descriptor', 'Domain', 'Domain Entity', 'Enumeration', 'Interchange', 'Inline Common', 'Shared Decimal', 'Shared Integer', 'Shared Short', 'Shared String', 'Subdomain', 'association', 'bool', 'choice', 'common', 'common extension', 'currency', 'date', 'decimal', 'descriptor', 'domain entity', 'duration', 'enumeration', 'inline common', 'integer', 'percent', 'shared decimal', 'shared integer', 'shared short', 'shared string', 'short', 'string', 'time', 'year', 'is queryable field', 'min value', 'max value', 'with context'}, column: 0, line: 9, token: TrailingText",
  "extraneous input 'TrailingText' expecting {'Abstract Entity', 'Association', 'End Namespace', 'Choice', 'Common', 'Descriptor', 'Domain', 'Domain Entity', 'Enumeration', 'Interchange', 'Inline Common', 'Shared Decimal', 'Shared Integer', 'Shared Short', 'Shared String', 'Subdomain', 'association', 'bool', 'choice', 'common', 'common extension', 'currency', 'date', 'decimal', 'descriptor', 'domain entity', 'duration', 'enumeration', 'inline common', 'integer', 'percent', 'shared decimal', 'shared integer', 'shared short', 'shared string', 'short', 'string', 'time', 'year', 'is queryable field', 'min value', 'max value', 'with context'}, column: 0, line: 9, token: TrailingText",
]
`;

exports[`when building choice with lowercase choice name should have extraneous input error 1`] = `
Array [
  "extraneous input 'entity' expecting ID, column: 9, line: 2, token: entity",
  "extraneous input 'entity' expecting ID, column: 9, line: 2, token: entity",
]
`;

exports[`when building choice with no choice name should have missing id error 1`] = `
Array [
  "missing ID at '[1]', column: 10, line: 2, token: [1]",
  "missing ID at '[1]', column: 10, line: 2, token: [1]",
]
`;

exports[`when building choice with no documentation should have mismatched input error 1`] = `
Array [
  "mismatched input 'integer' expecting 'documentation', column: 4, line: 3, token: integer",
  "mismatched input 'integer' expecting 'documentation', column: 4, line: 3, token: integer",
]
`;

exports[`when building choice with no property should have mismatched input error 1`] = `
Array [
  "mismatched input 'End Namespace' expecting {'association', 'bool', 'choice', 'common', 'common extension', 'currency', 'date', 'decimal', 'descriptor', 'domain entity', 'duration', 'enumeration', 'inline common', 'integer', 'percent', 'shared decimal', 'shared integer', 'shared short', 'shared string', 'short', 'string', 'time', 'year'}, column: 0, line: 5, token: End Namespace",
  "mismatched input 'End Namespace' expecting {'association', 'bool', 'choice', 'common', 'common extension', 'currency', 'date', 'decimal', 'descriptor', 'domain entity', 'duration', 'enumeration', 'inline common', 'integer', 'percent', 'shared decimal', 'shared integer', 'shared short', 'shared string', 'short', 'string', 'time', 'year'}, column: 0, line: 5, token: End Namespace",
]
`;

exports[`when building duplicate choices should have validation failures for each entity: when building duplicate choices should have validation failures for each entity -> Choice 1 message 1`] = `"Choice named EntityName is a duplicate declaration of that name."`;

exports[`when building duplicate choices should have validation failures for each entity: when building duplicate choices should have validation failures for each entity -> Choice 1 sourceMap 1`] = `
Object {
  "column": 2,
  "line": 9,
  "tokenText": "Choice",
}
`;

exports[`when building duplicate choices should have validation failures for each entity: when building duplicate choices should have validation failures for each entity -> Choice 2 message 1`] = `"Choice named EntityName is a duplicate declaration of that name."`;

exports[`when building duplicate choices should have validation failures for each entity: when building duplicate choices should have validation failures for each entity -> Choice 2 sourceMap 1`] = `
Object {
  "column": 2,
  "line": 2,
  "tokenText": "Choice",
}
`;
